File: demo/app/android/app/src/main/kotlin/walletsdk/flutter/converters/toFlutterTypes.kt
/*
Copyright Gen Digital Inc. All Rights Reserved.

SPDX-License-Identifier: Apache-2.0
*/

package walletsdk.flutter.converters

import dev.trustbloc.wallet.sdk.credential.InputDescriptor
import dev.trustbloc.wallet.sdk.credential.SubmissionRequirement
import dev.trustbloc.wallet.sdk.credential.SubmissionRequirementArray
import dev.trustbloc.wallet.sdk.display.Data
import dev.trustbloc.wallet.sdk.trustregistry.EvaluationResult
import dev.trustbloc.wallet.sdk.verifiable.CredentialsArray

public fun convertVerifiableCredentialsArray(arr: CredentialsArray): List<String> {
    return List(arr.length().toInt()
    ) { i: Int ->
        arr.atIndex(i.toLong()).serialize()
    }
}

public fun convertVerifiableCredentialsWithIdArray(arr: CredentialsArray): List<HashMap<String, String>> {
    return List(arr.length().toInt()
    ) { i: Int ->
        hashMapOf("id" to arr.atIndex(i.toLong()).id(), "content" to arr.atIndex(i.toLong()).serialize())
    }
}

public fun convertCredentialsDisplayDataArray(arr: Data): List<String> {
    return List(arr.credentialDisplaysLength().toInt()
    ) { i: Int ->
        arr.credentialDisplayAtIndex(i.toLong()).serialize()
    }
}

public fun convertSubmissionRequirementArray(requirements: SubmissionRequirementArray): List<HashMap<String, Any>> {
    return List(requirements.len().toInt()
    ) { i: Int ->
        convertSubmissionRequirement(requirements.atIndex(i.toLong()))
    }
}

private fun convertSubmissionRequirement(req: SubmissionRequirement): HashMap<String, Any> {
    val result = HashMap<String, Any>()
    result.put("rule", req.rule())
    result.put("name", req.name())
    result.put("min", req.min())
    result.put("max", req.max())
    result.put("count", req.count())
    result.put("nested", convertNestedSubmissionRequirement(req))
    result.put("inputDescriptors", convertInputDescriptors(req))
    return result;
}

private fun convertInputDescriptor(desc: InputDescriptor): HashMap<String, Any> {
    val matchedVCsID = List<String>(desc.matchedVCs.length().toInt()) { i: Int ->
        desc.matchedVCs.atIndex(i.toLong()).id()
    }

    val matchedVCs = List<String>(desc.matchedVCs.length().toInt()) { i: Int ->
        desc.matchedVCs.atIndex(i.toLong()).serialize()
    }

    return HashMap(mapOf(
            "id" to desc.id,
            "name" to desc.name,
            "purpose" to desc.purpose,
            "matchedVCsID" to matchedVCsID,
            "matchedVCs" to matchedVCs,
    ))
}

private fun convertNestedSubmissionRequirement(req: SubmissionRequirement): List<HashMap<String, Any>> {
    return List(req.nestedRequirementLength().toInt()
    ) { i: Int ->
        convertSubmissionRequirement(req.nestedRequirementAtIndex(i.toLong()))
    }
}

private fun convertInputDescriptors(req: SubmissionRequirement): List<HashMap<String, Any>> {
    return List(req.descriptorLen().toInt()
    ) { i: Int ->
        convertInputDescriptor(req.descriptorAtIndex(i.toLong()))
    }
}

public fun convertEvaluationResult(result: EvaluationResult): HashMap<String, Any> {
    val requestedAttestations = List(result.requestedAttestationLength().toInt()
    ) { i: Int ->
        result.requestedAttestationAtIndex(i.toLong())
    }

    return hashMapOf(
            "allowed" to result.allowed,
            "multipleCredentialAllowed" to  result.multipleCredentialAllowed,
            "errorCode" to result.errorCode,
            "errorMessage" to result.errorMessage,
            "requestedAttestations" to requestedAttestations,
            "denyReason" to result.denyReason(),
    )
}

File: demo/app/ios/Runner/coverters/toFlutterTypes.swift
/*
Copyright Gen Digital Inc. All Rights Reserved.

SPDX-License-Identifier: Apache-2.0
*/

import Foundation
import Walletsdk

func convertVerifiableCredentialsArray(arr: VerifiableCredentialsArray)-> Array<String> {
   var credList: [String] = []
   for i in 0..<arr.length() {
      credList.append((arr.atIndex(i)?.serialize(nil))!)
   }
    
   return credList
}

func convertEvaluationResult(res: TrustregistryEvaluationResult) -> Dictionary<String, Any> {
    var requestedAttestations: [String] = []
    for i in 0..<res.requestedAttestationLength() {
        requestedAttestations.append(res.requestedAttestation(at:i))
    }

    var resDic :[String: Any] = [
        "allowed":   res.allowed,
        "multipleCredentialAllowed": res.multipleCredentialAllowed,
        "errorCode":   res.errorCode,
        "errorMessage":   res.errorMessage,
        "requestedAttestations": requestedAttestations,
        "denyReason": res.denyReason()

    ]

    return resDic;
}

func convertVerifiableCredentialsWithIdArray(arr: VerifiableCredentialsArray)-> Array<Dictionary<String, Any>> {
   var credList: [Dictionary<String, Any>] = []
   for i in 0..<arr.length() {
       credList.append(["id": (arr.atIndex(i)?.id_())!, "content" :(arr.atIndex(i)?.serialize(nil))!])
   }
    
   return credList
}

func convertCredentialsDisplayDataArray(arr: DisplayData)-> Array<String> {
   var credList: [String] = []
    for i in 0..<arr.credentialDisplaysLength() {
      credList.append((arr.credentialDisplay(at: i)?.serialize(nil))!)
   }
    
   return credList
}

func convertSubmissionRequirementArray(requirements: CredentialSubmissionRequirementArray) -> Array<Dictionary<String, Any>> {
    var result: [Dictionary<String, Any>] = []
    for i in 0..<requirements.len() {
        result.append(convertSubmissionRequirement(req: requirements.atIndex(i)!))
    }
     
    return result
    
}

private func convertSubmissionRequirement(req: CredentialSubmissionRequirement)  -> Dictionary<String, Any> {
    
    var result : [String: Any] = [:]
    result["rule"] = req.rule()
    result["name"] = req.name()
    result["min"] = req.min()
    result["max"] = req.max()
    result["count"] = req.count()
    result["nested"] = convertNestedSubmissionRequirement(req: req)
    result["inputDescriptors"] = convertInputDescriptors(req: req)
    return result;
}

func convertInputDescriptor(desc: CredentialInputDescriptor) -> Dictionary<String, Any> {
    var matchedVCs: [String] = []
    var matchedVCsID: [String] = []
    
    for i in 0..<desc.matchedVCs!.length() {
        matchedVCs.append(desc.matchedVCs!.atIndex(i)!.serialize(nil))
    }

    for i in 0..<desc.matchedVCs!.length() {
        matchedVCsID.append(desc.matchedVCs!.atIndex(i)!.id_())
    }
    
    return [
        "id" : desc.id_,
        "name" : desc.name,
        "purpose" : desc.purpose,
        "matchedVCsID" : matchedVCsID,
        "matchedVCs" : matchedVCs,
    ]
}

func convertNestedSubmissionRequirement(req: CredentialSubmissionRequirement) -> Array<Dictionary<String, Any>> {
    var result: [Dictionary<String, Any>] = []

    for i in 0..<req.nestedRequirementLength() {
       result.append(convertSubmissionRequirement(req: req.nestedRequirement(at: i)!))
    }
    
    return result
}

func convertInputDescriptors(req: CredentialSubmissionRequirement)-> Array<Dictionary<String, Any>> {
    var result: [Dictionary<String, Any>] = []
    for i in 0..<req.descriptorLen() {
        result.append(convertInputDescriptor(desc: req.descriptor(at: i)!))
    }
    
    return result
}
